File: lib/java/openlinktoken/src/main/java/org/openlinktoken/crypto/CryptoSuite.java
/* SPDX-License-Identifier: MIT */
package org.openlinktoken.crypto;

import java.io.InvalidObjectException;
import java.io.Serializable;
import java.util.List;
import java.util.Map;

/**
 * Immutable contract for token primitives and exchange key establishment.
 */
public final class CryptoSuite implements Serializable {
    private static final long serialVersionUID = 1L;

    /** SHA-256 token digest identifier. */
    public static final String TOKEN_DIGEST_SHA256 = "SHA-256";

    /** SHA3-256 token digest identifier. */
    public static final String TOKEN_DIGEST_SHA3_256 = "SHA3-256";

    /** SHAKE256-256 token digest identifier. */
    public static final String TOKEN_DIGEST_SHAKE256_256 = "SHAKE256-256";

    /** HMAC-SHA256 token MAC identifier. */
    public static final String TOKEN_MAC_HS256 = "HS256";

    /** HMAC-SHA3-256 token MAC identifier. */
    public static final String TOKEN_MAC_HS3_256 = "HS3-256";

    /** KMAC256-256 token MAC identifier. */
    public static final String TOKEN_MAC_KMAC256_256 = "KMAC256-256";

    /** KMAC256 token MAC family prefix. */
    public static final String TOKEN_MAC_KMAC256_PREFIX = "KMAC256";

    /** AES-256-GCM token content-encryption identifier. */
    public static final String TOKEN_CONTENT_ENCRYPTION_A256GCM = "A256GCM";

    /** ECDH exchange key-agreement identifier. */
    public static final String EXCHANGE_KEY_AGREEMENT_ECDH = "ECDH";

    /** ML-KEM-768 exchange key-agreement identifier. */
    public static final String EXCHANGE_KEY_AGREEMENT_MLKEM768 = "ML-KEM-768";

    /** ECDH and ML-KEM-768 hybrid exchange key-agreement identifier. */
    public static final String EXCHANGE_KEY_AGREEMENT_ECDH_MLKEM768 = "ECDH+ML-KEM-768";

    /** ML-KEM exchange key-agreement family prefix. */
    public static final String EXCHANGE_KEY_AGREEMENT_MLKEM_PREFIX = "ML-KEM";

    /** The SHA-256 and ECDH crypto suite. */
    public static final CryptoSuite SUITE_SHA256_V1 =
            new CryptoSuite(
                    "suite-sha256-v1",
                    TOKEN_DIGEST_SHA256,
                    TOKEN_MAC_HS256,
                    TOKEN_CONTENT_ENCRYPTION_A256GCM,
                    EXCHANGE_KEY_AGREEMENT_ECDH,
                    1);

    /** The SHA-3-256 and ECDH crypto suite. */
    public static final CryptoSuite SUITE_SHA3_V1 =
            new CryptoSuite(
                    "suite-sha3-v1",
                    TOKEN_DIGEST_SHA3_256,
                    TOKEN_MAC_HS3_256,
                    TOKEN_CONTENT_ENCRYPTION_A256GCM,
                    EXCHANGE_KEY_AGREEMENT_ECDH,
                    1);

    /** The SHAKE256 and ML-KEM-768 crypto suite. */
    public static final CryptoSuite SUITE_PQ_SHAKE_V1 =
            new CryptoSuite(
                    "suite-pq-shake-v1",
                    TOKEN_DIGEST_SHAKE256_256,
                    TOKEN_MAC_KMAC256_256,
                    TOKEN_CONTENT_ENCRYPTION_A256GCM,
                    EXCHANGE_KEY_AGREEMENT_MLKEM768,
                    2);

    /** The SHA-3-256 and ML-KEM-768 crypto suite. */
    public static final CryptoSuite SUITE_PQ_V1 =
            new CryptoSuite(
                    "suite-pq-v1",
                    TOKEN_DIGEST_SHA3_256,
                    TOKEN_MAC_HS3_256,
                    TOKEN_CONTENT_ENCRYPTION_A256GCM,
                    EXCHANGE_KEY_AGREEMENT_MLKEM768,
                    2);

    /** The SHA-3-256, ECDH, and ML-KEM-768 hybrid crypto suite. */
    public static final CryptoSuite SUITE_PQ_HYBRID_V1 =
            new CryptoSuite(
                    "suite-pq-hybrid-v1",
                    TOKEN_DIGEST_SHA3_256,
                    TOKEN_MAC_HS3_256,
                    TOKEN_CONTENT_ENCRYPTION_A256GCM,
                    EXCHANGE_KEY_AGREEMENT_ECDH_MLKEM768,
                    2);

    private static final Map<String, CryptoSuite> REGISTRY = Map.of(
            SUITE_SHA256_V1.getSuiteId(),
            SUITE_SHA256_V1,
            SUITE_SHA3_V1.getSuiteId(),
            SUITE_SHA3_V1,
            SUITE_PQ_SHAKE_V1.getSuiteId(),
            SUITE_PQ_SHAKE_V1,
            SUITE_PQ_V1.getSuiteId(),
            SUITE_PQ_V1,
            SUITE_PQ_HYBRID_V1.getSuiteId(),
            SUITE_PQ_HYBRID_V1);

    private final String suiteId;
    private final String tokenDigestAlgorithm;
    private final String tokenMacAlgorithm;
    private final String tokenContentEncryption;
    private final String exchangeKeyAgreement;
    private final int exchangeConfigVersion;

    /**
     * Creates an immutable crypto suite definition.
     *
     * @param suiteId the stable suite identifier
     * @param tokenDigestAlgorithm the digest algorithm used for tokenization
     * @param tokenMacAlgorithm the keyed MAC algorithm used for token transformation
     * @param tokenContentEncryption the content-encryption algorithm used for match tokens
     * @param exchangeKeyAgreement the key-agreement mechanism used for exchanges
     * @param exchangeConfigVersion the exchange configuration version
     */
    CryptoSuite(
            String suiteId,
            String tokenDigestAlgorithm,
            String tokenMacAlgorithm,
            String tokenContentEncryption,
            String exchangeKeyAgreement,
            int exchangeConfigVersion) {
        this.suiteId = suiteId;
        this.tokenDigestAlgorithm = tokenDigestAlgorithm;
        this.tokenMacAlgorithm = tokenMacAlgorithm;
        this.tokenContentEncryption = tokenContentEncryption;
        this.exchangeKeyAgreement = exchangeKeyAgreement;
        this.exchangeConfigVersion = exchangeConfigVersion;
        validate();
    }

    /**
     * Validate the internal algorithm and exchange-version contract.
     *
     * @return this validated suite
     */
    public CryptoSuite validate() {
        if (!TOKEN_CONTENT_ENCRYPTION_A256GCM.equals(tokenContentEncryption)) {
            throw new IllegalArgumentException(
                    "Unsupported token content encryption '" + tokenContentEncryption + "'.");
        }
        if (exchangeConfigVersion == 1 && !EXCHANGE_KEY_AGREEMENT_ECDH.equals(exchangeKeyAgreement)) {
            throw new IllegalArgumentException("Exchange configuration version 1 only supports ECDH.");
        }
        if (exchangeConfigVersion == 2 && EXCHANGE_KEY_AGREEMENT_ECDH.equals(exchangeKeyAgreement)) {
            throw new IllegalArgumentException(
                    "Exchange configuration version 2 requires a non-ECDH key agreement.");
        }
        if (exchangeConfigVersion != 1 && exchangeConfigVersion != 2) {
            throw new IllegalArgumentException(
                    "Unsupported exchange configuration version '" + exchangeConfigVersion + "'.");
        }
        return this;
    }

    /**
     * Resolve a registered suite identifier.
     *
     * @param suiteId the suite identifier
     * @return the immutable suite definition
     */
    public static CryptoSuite fromId(String suiteId) {
        if (suiteId == null || suiteId.isBlank()) {
            throw new IllegalArgumentException("Crypto suite ID must be a non-empty string.");
        }
        CryptoSuite suite = REGISTRY.get(suiteId);
        if (suite == null) {
            throw new IllegalArgumentException("Unknown crypto suite '" + suiteId + "'. Supported suites: "
                    + String.join(", ", REGISTRY.keySet()) + ".");
        }
        return suite;
    }

    /**
     * Replaces a deserialized suite with its canonical registered instance.
     *
     * @return the registered suite matching this instance's identifier
     * @throws InvalidObjectException if this instance's identifier is not registered
     */
    private Object readResolve() throws InvalidObjectException {
        try {
            return fromId(suiteId);
        } catch (IllegalArgumentException exception) {
            InvalidObjectException invalidObjectException = new InvalidObjectException(exception.getMessage());
            invalidObjectException.initCause(exception);
            throw invalidObjectException;
        }
    }

    /**
     * Return the backward-compatible default suite.
     *
     * @return the default suite
     */
    public static CryptoSuite defaultSuite() {
        return SUITE_SHA256_V1;
    }

    /**
     * Return all registered suites.
     *
     * @return an immutable list of suites
     */
    public static List<CryptoSuite> all() {
        return List.copyOf(REGISTRY.values());
    }

    /**
     * Returns the stable identifier for this suite.
     *
     * @return the suite identifier
     */
    public String getSuiteId() {
        return suiteId;
    }

    /**
     * Returns the digest algorithm used to tokenize values.
     *
     * @return the token digest algorithm
     */
    public String getTokenDigestAlgorithm() {
        return tokenDigestAlgorithm;
    }

    /**
     * Returns the keyed MAC algorithm used to transform tokens.
     *
     * @return the token MAC algorithm
     */
    public String getTokenMacAlgorithm() {
        return tokenMacAlgorithm;
    }

    /**
     * Returns the content-encryption algorithm used for match tokens.
     *
     * @return the token content-encryption algorithm
     */
    public String getTokenContentEncryption() {
        return tokenContentEncryption;
    }

    /**
     * Returns the key-agreement mechanism used for exchanges.
     *
     * @return the exchange key-agreement mechanism
     */
    public String getExchangeKeyAgreement() {
        return exchangeKeyAgreement;
    }

    /**
     * Returns the exchange configuration version required by this suite.
     *
     * @return the exchange configuration version
     */
    public int getExchangeConfigVersion() {
        return exchangeConfigVersion;
    }

    /**
     * Indicates whether this suite uses a post-quantum key-agreement mechanism.
     *
     * @return {@code true} when the suite uses ML-KEM
     */
    public boolean isPostQuantum() {
        return exchangeKeyAgreement.contains(EXCHANGE_KEY_AGREEMENT_MLKEM_PREFIX);
    }
}

File: lib/python/openlinktoken/src/main/openlinktoken/crypto_suite.py
# SPDX-License-Identifier: MIT
"""Validated crypto-suite definitions shared by token and exchange consumers."""

from dataclasses import dataclass
from typing import ClassVar


class CryptoSuiteError(ValueError):
    """Raised when a crypto suite identifier or combination is not supported."""


@dataclass(frozen=True)
class CryptoSuite:
    """Immutable contract for token primitives and exchange key establishment."""

    TOKEN_DIGEST_SHA256: ClassVar[str] = "SHA-256"
    TOKEN_DIGEST_SHA3_256: ClassVar[str] = "SHA3-256"
    TOKEN_DIGEST_SHAKE256_256: ClassVar[str] = "SHAKE256-256"
    TOKEN_MAC_HS256: ClassVar[str] = "HS256"
    TOKEN_MAC_HS3_256: ClassVar[str] = "HS3-256"
    TOKEN_MAC_KMAC256_256: ClassVar[str] = "KMAC256-256"
    TOKEN_MAC_KMAC256_PREFIX: ClassVar[str] = "KMAC256"
    TOKEN_CONTENT_ENCRYPTION_A256GCM: ClassVar[str] = "A256GCM"
    EXCHANGE_KEY_AGREEMENT_ECDH: ClassVar[str] = "ECDH"
    EXCHANGE_KEY_AGREEMENT_MLKEM768: ClassVar[str] = "ML-KEM-768"
    EXCHANGE_KEY_AGREEMENT_ECDH_MLKEM768: ClassVar[str] = "ECDH+ML-KEM-768"
    EXCHANGE_KEY_AGREEMENT_MLKEM_PREFIX: ClassVar[str] = "ML-KEM"

    suite_id: str
    token_digest_algorithm: str
    token_mac_algorithm: str
    token_content_encryption: str
    exchange_key_agreement: str
    exchange_config_version: int

    SUITE_SHA256_V1: ClassVar["CryptoSuite"]
    SUITE_SHA3_V1: ClassVar["CryptoSuite"]
    SUITE_PQ_SHAKE_V1: ClassVar["CryptoSuite"]
    SUITE_PQ_V1: ClassVar["CryptoSuite"]
    SUITE_PQ_HYBRID_V1: ClassVar["CryptoSuite"]
    _REGISTRY: ClassVar[dict[str, "CryptoSuite"]]

    @classmethod
    def from_id(cls, suite_id: str) -> "CryptoSuite":
        """Resolve a registered suite identifier or raise a validation error."""
        if not isinstance(suite_id, str) or not suite_id.strip():
            raise CryptoSuiteError("Crypto suite ID must be a non-empty string.")

        try:
            return cls._REGISTRY[suite_id]
        except KeyError as error:
            supported = ", ".join(cls._REGISTRY)
            raise CryptoSuiteError(f"Unknown crypto suite '{suite_id}'. Supported suites: {supported}.") from error

    @classmethod
    def default(cls) -> "CryptoSuite":
        """Return the backward-compatible default suite."""
        return cls.SUITE_SHA256_V1

    @classmethod
    def all(cls) -> tuple["CryptoSuite", ...]:
        """Return all registered suites in stable identifier order."""
        return tuple(cls._REGISTRY.values())

    @property
    def is_post_quantum(self) -> bool:
        """Return whether the exchange agreement includes ML-KEM."""
        return self.EXCHANGE_KEY_AGREEMENT_MLKEM_PREFIX in self.exchange_key_agreement

    @property
    def minimum_mac_key_length(self) -> int:
        """Return the minimum hashing-secret length required by this suite's MAC."""
        return 32 if self.token_mac_algorithm.startswith(self.TOKEN_MAC_KMAC256_PREFIX) else 0

    def validate_hashing_secret(self, hashing_secret: bytes) -> bytes:
        """Validate and return hashing-secret bytes for this suite's MAC."""
        if not isinstance(hashing_secret, bytes):
            raise TypeError(f"Crypto suite '{self.suite_id}' requires hashing secret bytes.")

        minimum_length = self.minimum_mac_key_length
        if len(hashing_secret) < minimum_length:
            if self.token_mac_algorithm.startswith(self.TOKEN_MAC_KMAC256_PREFIX):
                raise CryptoSuiteError(
                    f"Crypto suite '{self.suite_id}': {self.TOKEN_MAC_KMAC256_PREFIX} requires a hashing secret "
                    f"of at least {minimum_length} bytes."
                )
            raise CryptoSuiteError(f"Crypto suite '{self.suite_id}' requires a non-empty hashing secret.")
        return hashing_secret

    def validate(self) -> "CryptoSuite":
        """Validate the suite's internal algorithm and version combination."""
        if self.token_content_encryption != self.TOKEN_CONTENT_ENCRYPTION_A256GCM:
            raise CryptoSuiteError(f"Unsupported token content encryption '{self.token_content_encryption}'.")
        if self.exchange_config_version == 1 and self.exchange_key_agreement != self.EXCHANGE_KEY_AGREEMENT_ECDH:
            raise CryptoSuiteError("Exchange configuration version 1 only supports ECDH.")
        if self.exchange_config_version == 2 and self.exchange_key_agreement == self.EXCHANGE_KEY_AGREEMENT_ECDH:
            raise CryptoSuiteError("Exchange configuration version 2 requires a non-ECDH key agreement.")
        if self.exchange_config_version not in {1, 2}:
            raise CryptoSuiteError(f"Unsupported exchange configuration version '{self.exchange_config_version}'.")
        return self


CryptoSuite.SUITE_SHA256_V1 = CryptoSuite(
    suite_id="suite-sha256-v1",
    token_digest_algorithm=CryptoSuite.TOKEN_DIGEST_SHA256,
    token_mac_algorithm=CryptoSuite.TOKEN_MAC_HS256,
    token_content_encryption=CryptoSuite.TOKEN_CONTENT_ENCRYPTION_A256GCM,
    exchange_key_agreement=CryptoSuite.EXCHANGE_KEY_AGREEMENT_ECDH,
    exchange_config_version=1,
)
CryptoSuite.SUITE_SHA3_V1 = CryptoSuite(
    suite_id="suite-sha3-v1",
    token_digest_algorithm=CryptoSuite.TOKEN_DIGEST_SHA3_256,
    token_mac_algorithm=CryptoSuite.TOKEN_MAC_HS3_256,
    token_content_encryption=CryptoSuite.TOKEN_CONTENT_ENCRYPTION_A256GCM,
    exchange_key_agreement=CryptoSuite.EXCHANGE_KEY_AGREEMENT_ECDH,
    exchange_config_version=1,
)
CryptoSuite.SUITE_PQ_SHAKE_V1 = CryptoSuite(
    suite_id="suite-pq-shake-v1",
    token_digest_algorithm=CryptoSuite.TOKEN_DIGEST_SHAKE256_256,
    token_mac_algorithm=CryptoSuite.TOKEN_MAC_KMAC256_256,
    token_content_encryption=CryptoSuite.TOKEN_CONTENT_ENCRYPTION_A256GCM,
    exchange_key_agreement=CryptoSuite.EXCHANGE_KEY_AGREEMENT_MLKEM768,
    exchange_config_version=2,
)
CryptoSuite.SUITE_PQ_V1 = CryptoSuite(
    suite_id="suite-pq-v1",
    token_digest_algorithm=CryptoSuite.TOKEN_DIGEST_SHA3_256,
    token_mac_algorithm=CryptoSuite.TOKEN_MAC_HS3_256,
    token_content_encryption=CryptoSuite.TOKEN_CONTENT_ENCRYPTION_A256GCM,
    exchange_key_agreement=CryptoSuite.EXCHANGE_KEY_AGREEMENT_MLKEM768,
    exchange_config_version=2,
)
CryptoSuite.SUITE_PQ_HYBRID_V1 = CryptoSuite(
    suite_id="suite-pq-hybrid-v1",
    token_digest_algorithm=CryptoSuite.TOKEN_DIGEST_SHA3_256,
    token_mac_algorithm=CryptoSuite.TOKEN_MAC_HS3_256,
    token_content_encryption=CryptoSuite.TOKEN_CONTENT_ENCRYPTION_A256GCM,
    exchange_key_agreement=CryptoSuite.EXCHANGE_KEY_AGREEMENT_ECDH_MLKEM768,
    exchange_config_version=2,
)

CryptoSuite._REGISTRY = {
    suite.suite_id: suite
    for suite in (
        CryptoSuite.SUITE_SHA256_V1,
        CryptoSuite.SUITE_SHA3_V1,
        CryptoSuite.SUITE_PQ_SHAKE_V1,
        CryptoSuite.SUITE_PQ_V1,
        CryptoSuite.SUITE_PQ_HYBRID_V1,
    )
}

for _suite in CryptoSuite._REGISTRY.values():
    _suite.validate()
